federated_aws_cli/cache.py

```diff
@@ -5,20 +5,50 @@
 import os
 import time
 
+from contextlib import contextmanager
 from hashlib import sha256
 from jose import jwt
 from stat import S_IRWXG, S_IRWXO, S_IRWXU
 
 
 # TODO: move to config
-CLOCK_SKEW_ALLOWANCE = 500  # 5 minutes
+CLOCK_SKEW_ALLOWANCE = 300         # 5 minutes
+GROUP_ROLE_MAP_CACHE_TIME = 3600   # 1 hour
+
+
 logger = logging.getLogger(__name__)
 
 # the cache directory is the same place we store the config
 cache_dir = os.path.join(os.path.expanduser("~"), ".federated_aws_cli")
 
 
-def __requires_safe_cache_dir(func):
+def _fix_permissions(path, permissions):
+    try:
+        os.chmod(path, permissions)
+
+        if _readable_by_others(path, fix=False):
+            logger.debug("Failed to repair permissions on: {}".format(path))
+        else:
+            logger.debug("Successfully repaired permissions on: {}".format(path))
+    except (IOError, PermissionError):
+        logger.debug("Failed to repair permissions on: {}".format(path))
+
+    return _readable_by_others(path, fix=False)
+
+
+def _readable_by_others(path, fix=True):
+    mode = os.stat(path).st_mode
+    readable_by_others = mode & S_IRWXG or mode & S_IRWXO
+
+    if readable_by_others and fix:
+        logger.debug("Cached file at {} has invalid permissions. Attempting to fix.".format(path))
+
+        return _fix_permissions(path, 0o600)
```

Comment:
This seems odd that `_readable_by_others` can return the output of `_fix_permissions` which can return the output of [`_readable_by_others`](https://github.com/mozilla-iam/federated-aws-cli/pull/84/files#diff-910d4774dffcbd9048bf2bba8a7b2eecR36).

I see that the `fix` argument governs this to prevent a loop.

It seems like calling [`os.chmod`](https://docs.python.org/3/library/os.html#os.chmod) will either succeed at changing permissions or it will raise an exception. Why [test if the file is readable by others after the chmod call](https://github.com/mozilla-iam/federated-aws-cli/pull/84/files#diff-910d4774dffcbd9048bf2bba8a7b2eecR29)?

This might be a simlpler approach that doesn't result in a need to govern that recursion with something like a `fix` argument

```
def _readable_by_others(path):
    mode = os.stat(path).st_mode
    readable_by_others = mode & S_IRWXG or mode & S_IRWXO

    if readable_by_others:
        logger.debug("Cached file at {} has invalid permissions. Attempting to fix.".format(path))
        permissions = 0o600
        try:
            os.chmod(path, permissions)
            return False
        except (IOError, PermissionError):
            logger.debug("Failed to repair permissions on: {}".format(path))
            return True
    else:
        return False
```

A variable and a function with the same name but one prefixed with `_` to be private feels confusing, maybe something like

```
def _readable_by_others(path):
    mode = os.stat(path).st_mode
    result = mode & S_IRWXG or mode & S_IRWXO

    if result:
        logger.debug("Cached file at {} has invalid permissions. Attempting to fix.".format(path))
        permissions = 0o600
        try:
            os.chmod(path, permissions)
            result = False
        except (IOError, PermissionError):
            logger.debug("Failed to repair permissions on: {}".format(path))
    return result
```